cmd/tpm-manager/tpm-manager.go

```diff
@@ -0,0 +1,544 @@
+		err = invalidateNode(node)
+		if err != nil {
+			return fmt.Errorf("Failed to invalidate node: %v", err)
+		}
+		return fmt.Errorf("Invalidating Node: Unable to obtain TPM quote for node %s: %v", address.String(), err)
+	}
+
+	err = tpm.ValidateLog(log, quote)
+	if err != nil {
+		err = invalidateNode(node)
+		if err != nil {
+			return fmt.Errorf("Failed to invalidate node: %v", err)
+		}
+		return fmt.Errorf("Invalidating Node: TPM event log does not match quote for node %s", address.String())
+	}
+
+	// Don't handle this error immediately - we want to update the annotations even if validation failed
+	logstate, err := tpm.ValidatePCRs(log, quote, manager.pcrConfigs)
+	jsonlog, jsonerr := json.Marshal(logstate)
+
+	if jsonerr == nil {
+		updateAnnotations(node, string(jsonlog))
+	} else {
+		glog.Errorf("Failed to serialise new log state: %v", jsonerr)
+	}
+
+	if err != nil {
+		err = invalidateNode(node)
+		if err != nil {
+			return fmt.Errorf("Failed to invalidate node: %v", err)
+		}
+		return fmt.Errorf("Invalidating Node: Unable to validate quote for node %s", address.String())
+	}
+
+	// If we've got this far then the node is trustworthy
+	err = trustNode(node)
+	if err != nil {
+		return fmt.Errorf("Failed to mark node as trustworthy")
+	}
+	return nil
+}
+
+func verifyAndUpdate(node *api.Node) {
+	err := verifyNode(node)
+	if err != nil {
+		glog.Errorf("Failed to verify node %s: %v", node.Name, err)
+	}
+	// The state that the node will be updated to
+	newstate := isTrusted(node)
+	newnode, err := manager.client.Nodes().Get(node.Name)
+	if err != nil {
+		glog.Errorf("Unable to obtain node state for %s: %v", node.Name, err)
+		return
+	}
+	if newnode.ObjectMeta.Annotations == nil {
+		newnode.ObjectMeta.Annotations = make(map[string]string)
+	}
+	currenttime := time.Now().Unix()
+
+	// If we're transitioning state, update the metadata
+	if isTrusted(newnode) != newstate {
+		if newstate == false {
+			newnode.ObjectMeta.Annotations[UntrustedSince] = strconv.FormatInt(currenttime, 10)
+			newnode.ObjectMeta.Annotations[TrustedSince] = ""
+		} else {
+			newnode.ObjectMeta.Annotations[UntrustedSince] = ""
+			newnode.ObjectMeta.Annotations[TrustedSince] = strconv.FormatInt(currenttime, 10)
+		}
+	}
+
+	newnode.ObjectMeta.Annotations[ValidationTime] = strconv.FormatInt(currenttime, 10)
+
+	// Ensure that the new node is tainted appropriately
+	if newstate == true {
+		trustNode(newnode)
+	} else {
+		invalidateNode(newnode)
+	}
+	newnode.ObjectMeta.Annotations[LogState] = node.ObjectMeta.Annotations[LogState]
+	newnode, err = manager.client.Nodes().Update(newnode)
+	if err != nil {
+		glog.Errorf("Unable to update node state for %s: %v", node.Name, err)
+		return
+	}
+}
+
+func verifyAllNodes() {
+	nodes, err := manager.client.Nodes().List(api.ListOptions{})
+	if err != nil {
+		glog.Errorf("Unable to obtain list of nodes")
+		return
+	}
+	for _, node := range nodes.Items {
+		verifyAndUpdate(&node)
+	}
+}
+
+func reverify() {
```

Comment:
The reverification here is at a configurable frequency, so if it's being triggered by the resync interval on the watch we'd be conflating two things. Verification can also take a few seconds, so we'd need to kick off another thread for the verification anyway rather than blocking here. But using the store rather than hitting the apiserver again seems like a pretty reasonable idea.